include/dsn/utility/join_point.h

```diff
@@ -24,454 +24,98 @@
-    }
-    bool put_replace(const char *base, advice_prototype fn, const char *name)
-    {
-        return join_point_base::put_replace(base, (void *)fn, name);
-    }
-
-    TReturn execute(T1 p1, T2 p2, TReturn default_return_value)
-    {
-        TReturn returnValue = default_return_value;
-        advice_entry *p = _hdr.next;
-        while (p != &_hdr) {
-            if (p->is_native) {
-                returnValue = (*(point_prototype *)&p->func)(p1, p2);
-            } else {
-                (*(advice_prototype *)&p->func)(p1, p2);
-            }
-            p = p->next;
-        }
-        return returnValue;
-    }
-};
+    using ReturnedAdviceT = R(Args...);
+    using AdviceT = void(Args...);
 
-template <typename T1, typename T2>
-class join_point<void, T1, T2, join_point_unused_type> : public join_point_base
-{
-public:
-    typedef void (*point_prototype)(T1, T2);
-    typedef void (*advice_prototype)(T1, T2);
+    // TODO(wutao): call it add_returned_advice()
+    void put_native(std::function<ReturnedAdviceT> fn) { _ret_advice_entries.push_front(fn); }
 
-public:
-    join_point(const char *name) : join_point_base(name) {}
-    bool put_native(point_prototype point)
-    {
-        return join_point_base::put_front((void *)point, "native", true);
-    }
-    bool put_front(advice_prototype fn, const char *name)
-    {
-        return join_point_base::put_front((void *)fn, name);
-    }
-    bool put_back(advice_prototype fn, const char *name)
-    {
-        return join_point_base::put_back((void *)fn, name);
-    }
-    bool put_before(const char *base, advice_prototype fn, const char *name)
-    {
-        return join_point_base::put_before(base, (void *)fn, name);
-    }
-    bool put_after(const char *base, advice_prototype fn, const char *name)
+    // TODO(wutao): call it add_advice()
+    void put_back(std::function<AdviceT> fn, const char * /*unused*/)
     {
-        return join_point_base::put_after(base, (void *)fn, name);
-    }
-    bool put_replace(const char *base, advice_prototype fn, const char *name)
-    {
-        return join_point_base::put_replace(base, (void *)fn, name);
+        _advice_entries.push_back(std::move(fn));
     }
 
-    void execute(T1 p1, T2 p2)
+    void put_front(std::function<AdviceT> fn, const char * /*unused*/)
     {
-        advice_entry *p = _hdr.next;
-        while (p != &_hdr) {
-            if (p->is_native) {
-                (*(point_prototype *)&p->func)(p1, p2);
-            } else {
-                (*(advice_prototype *)&p->func)(p1, p2);
-            }
-            p = p->next;
-        }
+        _advice_entries.push_front(std::move(fn));
     }
-};
 
-template <typename TReturn, typename T1>
-class join_point<TReturn, T1, join_point_unused_type, join_point_unused_type>
-    : public join_point_base
-{
-public:
-    typedef TReturn (*point_prototype)(T1);
-    typedef void (*advice_prototype)(T1);
+    const char *name() const { return _name.c_str(); }
```

Comment:
```
const std::string& name() const { return _name; }
```

Reply:
this is what it used to be. I don't wanna change any API.